### PR TITLE
Add scraper for Rekord Pretoria-North

### PR DESCRIPTION
**Pull request by schalkneethling**
@jbothma This seems to be working. Two questions though:

1. Some urls seem to still get scraped even though they are in the `deny` list of the `link_extractor`. Is there a specific reason for this?
2. Out of curiosity, where does the data go that is scraped? It seems to only output to standard out.

r?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -129,7 +129,7 @@ def parse(self, response):
     ...
 ```
 
-#### Using css instead of xpath for classes
+#### Using xpath css instead of css for classes
```
Comment (jbothma) [inline README.md]: Looks like the intent of the heading wasn't clear - does this make more sense?
Primarily this section is about how css is better than xpath for classes. Secondly it tries to confirm that yes, both are options.

```suggestion
#### CSS lookup is preferred over XPath for HTML classes
```

Comment (jbothma) [inline README.md]: Looks like the intent of the heading wasn't clear - does this make more sense?
Primarily this section is about how css is better than xpath for classes. Secondly it tries to confirm that yes, both are options.

```suggestion
#### CSS lookup is preferred over XPath for HTML classes
```

**Comment by jbothma** [comment]
>     1. Some urls seem to still get scraped even though they are in the `deny` list of the `link_extractor`. Is there a specific reason for this?

Good question - I'm not sure - I suspect they might be using "OR" on the list of deny rules instead of "AND" - will check up on this. I suspect we also have to pass allowed_domains into the LinkExtractor
 
>     2. Out of curiosity, where does the data go that is scraped? It seems to only output to standard out.

When I deploy it I configure scrapenews.pipelines.AlephPipeline which pushes the news into a v1 instance of https://github.com/alephdata - that's what gives is the full text search.

**Comment by schalkneethling** [comment]
Hey @jbothma, anything you need from me to get this one merged? Thanks!

**Comment by jbothma** [comment]
Not sure - Sorry I haven't gotten to it - work bleeds into free time too
much this last two weeks - will get onto it ASAP

JD

File: scrapenews/spiders/rekordnorth.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+import pytz
+
+SAST = pytz.timezone('Africa/Johannesburg')
+
+class dfaSpider(CrawlSpider):
+    name = 'rekordnorth'
+    allowed_domains = ['rekordnorth.co.za']
+    start_urls = ['https://rekordnorth.co.za/category/news-headlines/local-news/']
+
+    link_extractor = LinkExtractor(
```
Comment (jbothma) [inline scrapenews/spiders/rekordnorth.py]: try this - it seems to work fine - I was right - we should pass allowed_domains into LinkExtractor's allow_domains (note the different spelling :/ )

I noticed that the articles on this site all seem to have an integer as the first URL "part" - so it's easier to whitelist than blacklist.

```diff
diff --git a/scrapenews/spiders/rekordnorth.py b/scrapenews/spiders/rekordnorth.py
index b008475..8a86a81 100644
--- a/scrapenews/spiders/rekordnorth.py
+++ b/scrapenews/spiders/rekordnorth.py
@@ -14,70 +14,8 @@ class dfaSpider(CrawlSpider):
     start_urls = ['https://rekordnorth.co.za/category/news-headlines/local-news/']
 
     link_extractor = LinkExtractor(
-        allow=('https://rekordnorth.co.za', ),
-        deny=(
-            'rekordnorth.co.za/pretoria-north-news-team',
-            'rekordnorth.co.za/epapers',
-            'rekordnorth.co.za/feat/title-deeds/',
-            'rekordnorth.co.za/advertise-online/',
-            'rekordnorth.co.za/category/featured-content/',
-            'rekordnorth.co.za/beauty-health-all-year-round/',
-            'rekordnorth.co.za/rss-output/',
-            'rekordnorth.co.za/all4women',
-            'rekordnorth.co.za/national-news/',
-            'rekordnorth.co.za/local-motoring-news/',
-            'rekordnorth.co.za/category/image-gallery/',
-            'rekordnorth.co.za/category/image-gallery/general-news/',
-            'rekordnorth.co.za/category/image-gallery/community/',
-            'rekordnorth.co.za/category/image-gallery/sport-news/',
-            'rekordnorth.co.za/category/image-gallery/school-news/',
-            'rekordnorth.co.za/video-gallery/',
-            'rekordnorth.co.za/category/opinion-edition/letters/',
-            'rekordnorth.co.za/category/opinion-edition/blogs/',
-            'rekordnorth.co.za/category/opinion-edition/blogs/simply-delicious/',
-            'rekordnorth.co.za/category/lifestyle-news/competitions-in-pretoria-north/',
-            'rekordnorth.co.za/category/lifestyle-news/entertainment-news/',
-            'rekordnorth.co.za/events/event/',
-            'rekordnorth.co.za/category/news-headlines/',
-            'rekordnorth.co.za/category/sports-news/',
-            'rekordnorth.co.za/category/opinion-edition/',
-            'rekordnorth.co.za/category/lifestyle-news/',
-            'rekordnorth.co.za/online-classifieds/',
-            'rekordnorth.co.za/place-ad/',
-            'rekordnorth.co.za/terms-and-conditions/',
-            'rekordnorth.co.za/privacy-policy/',
-            'rekordnorth.co.za/builders/',
-            'rekordnorth.co.za/property/for-sale/',
-            'rekordnorth.co.za/property/to-rent/',
-            'rekordnorth.co.za/rest-assured/',
-            'rekordeast.co.za/parenting/',
-            'rekordnorth.co.za/support-local-business/',
-            'rekordnorth.co.za/i-love-my-city/',
-            'rekordnorth.co.za/?p=128157',
-            'rekordnorth.co.za/student-living/',
-            'rekordnorth.co.za/be-your-best/',
-            'rekordnorth.co.za/magalieskruin-centre/',
-            'rekordnorth.co.za/about-us',
-            '//biz.rekordnorth.co.za/',
-            '//www.guzzle.co.za/',
-            '//www.autodealer.co.za/',
-            '//iab.com/',
-            '//www.bestofpretoria.co.za/',
-            '//localnewsnetwork.co.za/',
-            '//www.caxton.co.za/',
-            '//facebook.com/',
-            '//twitter.com/',
-            '//southcoastherald.co.za/',
-            '//rekordcenturion.co.za/',
-            '//kemptonexpress.co.za/',
-            '//northglennews.co.za/',
-            '//albertonrecord.co.za/',
-            '//ladysmithgazette.co.za/',
-            '//witbanknews.co.za/',
-            '//roodepoortrecord.co.za/',
-            '//southerncourier.co.za/',
-            '//krugersdorpnews.co.za/',
-        )
+        allow=('https://rekordnorth.co.za/\d+/', ),
+        allow_domains=allowed_domains
     )
 
     rules = (
```

File: scrapenews/spiders/rekordnorth.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+        canonical_url = response.xpath('//link[@rel="canonical"]/@href').extract_first()
+        title = response.xpath('//h1[@class="entry-title"]/text()').extract_first()
+        self.logger.info('%s %s', response.url, title)
+        # should we be using canonical_url instead of response.url for the above?
+        og_type = response.xpath('//meta[@property="og:type"]/@content').extract_first()
+        if og_type == 'article':
+            article_body = response.css('div.entry-content')
+            body_html = " ".join(article_body.xpath('//p').extract())
+            byline = response.css('div.author-name').css('::text').extract()
```
Comment (jbothma) [inline scrapenews/spiders/rekordnorth.py]: this should be a string, not an array of strings - you could probably just change it to `extract_first()` - perhaps I should add some validation to give better feedback to contributors.

File: scrapenews/spiders/rekordnorth.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    def parse_item(self, response):
+
+        canonical_url = response.xpath('//link[@rel="canonical"]/@href').extract_first()
+        title = response.xpath('//h1[@class="entry-title"]/text()').extract_first()
+        self.logger.info('%s %s', response.url, title)
+        # should we be using canonical_url instead of response.url for the above?
+        og_type = response.xpath('//meta[@property="og:type"]/@content').extract_first()
+        if og_type == 'article':
+            article_body = response.css('div.entry-content')
+            body_html = " ".join(article_body.xpath('//p').extract())
```
Comment (jbothma) [inline scrapenews/spiders/rekordnorth.py]: this should be a string, not an array of strings - you could probably just change it to extract_first()

**Comment by jbothma** [comment]
I sommer quickly made these changes in https://github.com/public-people/scrape-news/pull/20 and added a whitelist for the news articles based on the category "tags" at the top.

Thanks so much for the contribution!

**Pull request merged**